File: scripts/persist-consensus-artifacts.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";
import { parseReviewerIds } from "./shared/reviewers-json.mjs";

function writeTextFile(filePath, value) {
  const normalized = String(value ?? "");
  fs.writeFileSync(filePath, `${normalized}\n`, "utf8");
}

function copyOrCreateEmptyFile(sourcePath, targetPath) {
  if (fs.existsSync(sourcePath)) {
    fs.copyFileSync(sourcePath, targetPath);
    return;
  }
  writeTextFile(targetPath, "");
}

export function persistConsensusArtifacts({
  runnerTemp,
  reviewersJson,
  consensusReports,
  outcome,
  openFindingsCount,
  reviewerErrorsCount,
  commentPath,
  ledgerPath,
}) {
  const normalizedRunnerTemp = String(runnerTemp || "").trim();
  if (!normalizedRunnerTemp) {
    throw new Error("RUNNER_TEMP is required");
  }

  const normalizedCommentPath = String(commentPath || "").trim();
  if (!normalizedCommentPath) {
    throw new Error("COMMENT_PATH is required");
  }
  const normalizedLedgerPath = String(ledgerPath || "").trim();
  if (!normalizedLedgerPath) {
    throw new Error("LEDGER_PATH is required");
  }

  const reviewerIds = parseReviewerIds(reviewersJson);
  const sourceReportsDir = path.join(normalizedRunnerTemp, "lgtm-reports");
  const targetDir = path.join(normalizedRunnerTemp, "lgtm");
  fs.mkdirSync(targetDir, { recursive: true });

  for (const reviewerId of reviewerIds) {
    const sourcePath = path.join(sourceReportsDir, `${reviewerId}.json`);
    const targetPath = path.join(targetDir, `${reviewerId}.json`);
    copyOrCreateEmptyFile(sourcePath, targetPath);
  }

  writeTextFile(path.join(targetDir, "reports-merged.json"), consensusReports);
  writeTextFile(path.join(targetDir, "outcome.txt"), outcome);
  writeTextFile(path.join(targetDir, "open-findings-count.txt"), openFindingsCount);
  // Backward-compat artifact path retained for external consumers.
  writeTextFile(path.join(targetDir, "blocking-findings-count.txt"), openFindingsCount);
  writeTextFile(path.join(targetDir, "reviewer-errors-count.txt"), reviewerErrorsCount);
  fs.copyFileSync(normalizedCommentPath, path.join(targetDir, "comment.md"));
  fs.copyFileSync(normalizedLedgerPath, path.join(targetDir, "findings-ledger.json"));
}

function main() {
  persistConsensusArtifacts({
    runnerTemp: process.env.RUNNER_TEMP,
    reviewersJson: process.env.REVIEWERS_JSON,
    consensusReports: process.env.CONSENSUS_REPORTS,
    outcome: process.env.OUTCOME,
    openFindingsCount: process.env.OPEN_FINDINGS_COUNT || process.env.BLOCKING_FINDINGS_COUNT,
    reviewerErrorsCount: process.env.REVIEWER_ERRORS_COUNT,
    commentPath: process.env.COMMENT_PATH,
    ledgerPath: process.env.LEDGER_PATH,
  });
}

function isCliMain() {
  if (!process.argv[1]) return false;
  return import.meta.url === pathToFileURL(process.argv[1]).href;
}

if (isCliMain()) {
  try {
    main();
  } catch (error) {
    console.error(error.message);
    process.exit(1);
  }
}

File: scripts/persist_consensus_artifacts.py
from __future__ import annotations

import os
from pathlib import Path

from scripts.shared.reviewers_json import parse_reviewer_ids


def write_text_file(file_path: Path, value: str) -> None:
    file_path.write_text(f"{value}\n", encoding="utf-8")


def copy_or_create_empty_file(source_path: Path, target_path: Path) -> None:
    if source_path.exists():
        target_path.write_bytes(source_path.read_bytes())
        return

    write_text_file(target_path, "")


def persist_consensus_artifacts(
    *,
    runner_temp: str,
    reviewers_json: str,
    consensus_reports: str,
    outcome: str,
    open_findings_count: str,
    reviewer_errors_count: str,
    comment_path: str,
    ledger_path: str,
) -> None:
    normalized_runner_temp = runner_temp.strip()
    if not normalized_runner_temp:
        raise ValueError("RUNNER_TEMP is required")

    normalized_comment_path = comment_path.strip()
    if not normalized_comment_path:
        raise ValueError("COMMENT_PATH is required")

    normalized_ledger_path = ledger_path.strip()
    if not normalized_ledger_path:
        raise ValueError("LEDGER_PATH is required")

    reviewer_ids = parse_reviewer_ids(reviewers_json)
    source_reports_dir = Path(normalized_runner_temp) / "lgtm-reports"
    target_dir = Path(normalized_runner_temp) / "lgtm"
    target_dir.mkdir(parents=True, exist_ok=True)

    for reviewer_id in reviewer_ids:
        source_path = source_reports_dir / f"{reviewer_id}.json"
        target_path = target_dir / f"{reviewer_id}.json"
        copy_or_create_empty_file(source_path, target_path)

    write_text_file(target_dir / "reports-merged.json", consensus_reports)
    write_text_file(target_dir / "outcome.txt", outcome)
    write_text_file(target_dir / "open-findings-count.txt", open_findings_count)
    write_text_file(target_dir / "reviewer-errors-count.txt", reviewer_errors_count)

    (target_dir / "comment.md").write_bytes(Path(normalized_comment_path).read_bytes())
    (target_dir / "findings-ledger.json").write_bytes(Path(normalized_ledger_path).read_bytes())


def main() -> None:
    persist_consensus_artifacts(
        runner_temp=os.getenv("RUNNER_TEMP", ""),
        reviewers_json=os.getenv("REVIEWERS_JSON", "[]"),
        consensus_reports=os.getenv("CONSENSUS_REPORTS", ""),
        outcome=os.getenv("OUTCOME", ""),
        open_findings_count=os.getenv("OPEN_FINDINGS_COUNT", "0"),
        reviewer_errors_count=os.getenv("REVIEWER_ERRORS_COUNT", "0"),
        comment_path=os.getenv("COMMENT_PATH", ""),
        ledger_path=os.getenv("LEDGER_PATH", ""),
    )


if __name__ == "__main__":
    main()
